apps/emqx_plugin_libs/src/emqx_slow_subs/emqx_slow_subs.erl

```diff
@@ -0,0 +1,308 @@
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_slow_subs).
+
+-behaviour(gen_server).
+
+-include_lib("include/emqx.hrl").
+-include_lib("include/logger.hrl").
+-include_lib("emqx_plugin_libs/include/emqx_slow_subs.hrl").
+
+-logger_header("[SLOW Subs]").
+
+-export([ start_link/1, on_stats_update/2, enable/0
+        , disable/0, clear_history/0, init_topk_tab/0
+        ]).
+
+%% gen_server callbacks
+-export([ init/1
+        , handle_call/3
+        , handle_cast/2
+        , handle_info/2
+        , terminate/2
+        , code_change/3
+        ]).
+
+-compile(nowarn_unused_type).
+
+-type state() :: #{ config := proplist:proplist()
+                  , enable := boolean()
+                  , last_tick_at := pos_integer()
+                  }.
+
+-type log() :: #{ rank := pos_integer()
+                , clientid := emqx_types:clientid()
+                , latency := float()
+                , type := emqx_message_latency_stats:latency_type()
+                }.
+
+-type window_log() :: #{ last_tick_at := pos_integer()
+                       , logs := [log()]
+                       }.
+
+-type message() :: #message{}.
+
+-import(proplists, [get_value/2]).
+
+-type stats_update_args() :: #{ clientid := emqx_types:clientid()
+                              , latency := float()
+                              , type := emqx_message_latency_stats:latency_type()
+                              , last_insert_value := non_neg_integer()
+                              , update_time := pos_integer()
+                              }.
+
+-type stats_update_env() :: #{max_size := pos_integer()}.
+
+-ifdef(TEST).
+-define(EXPIRE_CHECK_INTERVAL, timer:seconds(1)).
+-else.
+-define(EXPIRE_CHECK_INTERVAL, timer:seconds(10)).
+-endif.
+
+-define(NOW, erlang:system_time(millisecond)).
+-define(NOTICE_TOPIC_NAME, "slow_subs").
+-define(DEF_CALL_TIMEOUT, timer:seconds(10)).
+
+%% erlang term order
+%% number < atom < reference < fun < port < pid < tuple < list < bit string
+
+%% ets ordered_set is ascending by term order
+
+%%--------------------------------------------------------------------
+%%--------------------------------------------------------------------
+%% APIs
+%%--------------------------------------------------------------------
+%% @doc Start the st_statistics
+-spec(start_link(Env :: list()) -> emqx_types:startlink_ret()).
+start_link(Env) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Env], []).
+
+-spec on_stats_update(stats_update_args(), stats_update_env()) -> true.
+on_stats_update(#{clientid := ClientId,
+                  latency := Latency,
+                  type := Type,
+                  last_insert_value := LIV,
+                  update_time := Ts},
+                #{max_size := MaxSize}) ->
+
+    Index = ?INDEX(Latency, ClientId),
+
+    case ets:info(?TOPK_TAB, size) of
```

Comment:
this will be a very expensive call, add a note here and we could improve it later.

Reply:
BTW, I like the old method of using counters module.